lib/TPP/Transforms/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,179 @@
+//===CombineXsmmPass.cpp --------------------------------------*----C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM
+// Exceptions. / See https://llvm.org/LICENSE.txt for license information. /
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Passes.h"
+#include "TPP/Transforms/Utils/BuilderUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+using namespace mlir;
+
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Transforms/Utils/VNNIUtils.h"
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_COMBINEXSMMOPPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+namespace {
+
+static FailureOr<DenseI64ArrayAttr>
+getSizesAndLeadingDimForBrgemmOp(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+
+  auto memrefA = opTy.getOperand(1).getType();
+  auto memrefB = opTy.getOperand(2).getType();
+  auto memrefC = opTy.getOperand(3).getType();
+
+  if (!isa<ShapedType>(memrefC) || !isa<ShapedType>(memrefA)) {
+    return failure();
+  }
+  int64_t m, n, k;
+  m = memrefC.cast<ShapedType>().getShape()[0];
+  n = memrefC.cast<ShapedType>().getShape()[1];
+  k = memrefA.cast<ShapedType>().getShape()[2];
+
+  auto ldaDim = xsmm::utils::getLeadingDim(memrefA, /*pos=*/1);
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb =
+      (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::BRGEMM_INS,
+                                   memrefB.cast<MemRefType>()))
+          ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+          : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+static ArrayAttr getBrgemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
```

Comment:
Why not having this method in TableGen in the op?

Reply:
This will be done in subsequent iterations, I tried moving it but kept encountering compilation errors.